File: ds/stack.py
from __future__ import annotations


class Node:
    def __init__(self, value, next=None) -> None:
        self.val = value
        self.next = next


class Stack:
    def __init__(self) -> None:
        self.head = None
        self.size = 0

    def empty(self) -> bool:
        """Checks if stack is empty"""
        return self.size == 0 and self.head is None

    def peek(self):
        """Look at the top element without removing it"""
        if self.head is None:
            return None
        return self.head.val

    def pop(self):
        """Remove the top element and return that objects value"""
        self.size = max(0, self.size - 1)
        if self.head is None:
            return None

        if self.size == 0:
            tmp = self.head
            self.head = None
            return tmp.val

        node = self.head
        self.head = node.next
        return node.val

    def push(self, value):
        """Push an element to the top of the stack"""
        node = Node(value)
        self.size += 1

        node.next = self.head
        self.head = node

    def search(self, value):
        """Return the index of the value."""
        index = 1
        curr = self.head
        while curr:
            print(curr.val)
            if curr.val == value:
                return index
            curr = curr.next
            index += 1
        return -1

File: go/structures/stack.go
package structures

import "errors"

// Stack is an interface representing a Last-In-First-Out (LIFO) data structure.
type Stack[T any] interface {
	// Push adds an element to the top of the stack.
	Push(item T)
	// Pop removes and returns the top element of the stack.
	// Returns an error if the stack is empty.
	Pop() (T, error)
	// Peek returns the top element of the stack without removing it.
	// Returns an error if the stack is empty.
	Peek() (T, error)
	// IsEmpty returns true if the stack is empty, false otherwise.
	IsEmpty() bool
	// Size returns the number of elements in the stack.
	Size() int
	// Clear removes all elements from the stack.
	Clear()
}

var ErrEmptyStack = errors.New("stack is empty")

type AnyNode[T any] struct {
	value T
	next  *AnyNode[T]
}

type ListStack[T any] struct {
	head *AnyNode[T]
	size int
}

func (s *ListStack[T]) Push(item T) {
	s.size++
	node := &AnyNode[T]{value: item}
	if s.head == nil {
		s.head = node
	}
}

func (s *ListStack[T]) Pop() (T, error) {
	if s.head == nil {
		var zero T
		return zero, ErrEmptyStack
	}

	top := s.head
	s.head = top.next
	s.size--
	return top.value, nil
}

func (s *ListStack[T]) Peek() (T, error) {
	if s.head == nil {
		var zero T
		return zero, ErrEmptyStack
	}

	return s.head.value, nil
}

func (s *ListStack[T]) Size() int {
	return s.size
}

func (s *ListStack[T]) Clear() {
	s.size = 0
	s.head = nil
}
